transports/quic/src/muxer.rs

```diff
@@ -0,0 +1,360 @@
+            })),
+        }
+    }
+}
+
+impl StreamMuxer for QuicMuxer {
+    type Substream = Substream;
+    type Error = ConnectionError;
+
+    fn poll(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<StreamMuxerEvent, Self::Error>> {
+        let mut inner = self.inner.lock();
+        // Poll the inner [`quinn_proto::Connection`] for events and wake
+        // the wakers of related poll-based methods.
+        while let Poll::Ready(event) = inner.connection.poll_event(cx) {
+            match event {
+                ConnectionEvent::Connected | ConnectionEvent::HandshakeDataReady => {
+                    debug_assert!(
+                        false,
+                        "Unexpected event {:?} on established QUIC connection",
+                        event
+                    );
+                }
+                ConnectionEvent::ConnectionLost(err) => return Poll::Ready(Err(err)),
+                ConnectionEvent::StreamOpened => {
+                    if let Some(waker) = inner.poll_outbound_waker.take() {
+                        waker.wake();
+                    }
+                }
+                ConnectionEvent::StreamReadable(substream) => {
+                    if let Some(substream) = inner.substreams.get_mut(&substream) {
+                        if let Some(waker) = substream.read_waker.take() {
+                            waker.wake();
+                        }
+                    }
+                }
+                ConnectionEvent::StreamWritable(substream) => {
+                    if let Some(substream) = inner.substreams.get_mut(&substream) {
+                        if let Some(waker) = substream.write_waker.take() {
+                            waker.wake();
+                        }
+                    }
+                }
+                ConnectionEvent::StreamFinished(substream)
+                | ConnectionEvent::StreamStopped(substream) => {
+                    if let Some(substream) = inner.substreams.get_mut(&substream) {
+                        if let Some(waker) = substream.read_waker.take() {
+                            waker.wake();
+                        }
+                        if let Some(waker) = substream.write_waker.take() {
+                            waker.wake();
+                        }
+                        if let Some(waker) = substream.finished_waker.take() {
+                            waker.wake();
+                        }
+                    }
+                }
+                ConnectionEvent::StreamAvailable => {
+                    if let Some(waker) = inner.poll_inbound_waker.take() {
+                        waker.wake();
+                    }
+                }
+            }
+        }
+        inner.poll_connection_waker = Some(cx.waker().clone());
+
+        // TODO: If connection migration is enabled (currently disabled) address
+        // change on the connection needs to be handled.
+
+        Poll::Pending
```

Comment:
Putting waker + `Pending` together makes this easier to reason about IMO.

```suggestion
        // TODO: If connection migration is enabled (currently disabled) address
        // change on the connection needs to be handled.

        inner.poll_connection_waker = Some(cx.waker().clone());
        Poll::Pending
```